File: client/typescript-client/src/validation.ts
import { ApiValueError } from "./exceptions";
import {
  EXCHANGE_SEGMENT,
  PRODUCT,
  ORDER_TYPE,
  SEGMENT_LIMITS,
  EXCHANGE_LIMITS,
  PRODUCT_LIMITS,
} from "./settings";

function nonEmpty(val: unknown, name: string): void {
  if (val === undefined || val === null || String(val).trim() === "") {
    throw new ApiValueError(`${name} is mandatory`);
  }
}

const VALIDITY_VALUES = ["DAY", "IOC"];
const PLACE_ORDER_TT = ["B", "S", "Buy", "Sell"];
const MARGIN_TT = ["B", "S", "Buy", "Sell", "sell", "buy"];

export function validatePlaceOrder(p: {
  exchange_segment: string;
  product: string;
  price: string;
  order_type: string;
  quantity: string;
  validity: string;
  trading_symbol: string;
  transaction_type: string;
}): void {
  Object.entries(p).forEach(([k, v]) => nonEmpty(v, k));
  if (!(p.exchange_segment in EXCHANGE_SEGMENT))
    throw new ApiValueError(`invalid exchange_segment: ${p.exchange_segment}`);
  if (!(p.product in PRODUCT)) throw new ApiValueError(`invalid product: ${p.product}`);
  if (!(p.order_type in ORDER_TYPE))
    throw new ApiValueError(`invalid order_type: ${p.order_type}`);
  if (!VALIDITY_VALUES.includes(p.validity))
    throw new ApiValueError("Invalid validity. Allowed values are DAY, IOC.");
  if (!PLACE_ORDER_TT.includes(p.transaction_type))
    throw new ApiValueError("Invalid transaction type. Allowed values are B or Buy, S or Sell.");
}

export function validateCancelOrder(orderId: string): void {
  nonEmpty(orderId, "order_id");
}

export function validateOrderHistory(orderId: string): void {
  nonEmpty(orderId, "order_id");
}

export function validateMargin(p: {
  exchange_segment: string;
  price: string;
  order_type: string;
  product: string;
  quantity: string;
  instrument_token: string;
  transaction_type: string;
}): void {
  Object.entries(p).forEach(([k, v]) => nonEmpty(v, k));
  if (!(p.exchange_segment in EXCHANGE_SEGMENT))
    throw new ApiValueError(`invalid exchange_segment: ${p.exchange_segment}`);
  if (!(p.product in PRODUCT))
    throw new ApiValueError(`invalid product: ${p.product}`);
  if (!(p.order_type in ORDER_TYPE))
    throw new ApiValueError(`invalid order_type: ${p.order_type}`);
  if (!MARGIN_TT.includes(p.transaction_type))
    throw new ApiValueError("Invalid transaction type. Allowed values are B or Buy, S or Sell.");
}

export function validateLimits(segment: string, exchange: string, product: string): void {
  if (!SEGMENT_LIMITS.includes(segment))
    throw new ApiValueError(`invalid segment: ${segment}`);
  if (!EXCHANGE_LIMITS.includes(exchange))
    throw new ApiValueError(`invalid exchange: ${exchange}`);
  if (!PRODUCT_LIMITS.includes(product))
    throw new ApiValueError(`invalid product: ${product}`);
}

File: client/go-client/neoapi/validation.go
package neoapi

import (
	"fmt"
	"strings"
)

// Response is the loose response type — mirrors Python dict.
type Response map[string]any

func nonEmpty(val, name string) error {
	if strings.TrimSpace(val) == "" {
		return &APIValueError{Msg: fmt.Sprintf("%s is mandatory", name)}
	}
	return nil
}

// ValidatePlaceOrder mirrors req_data_validation.place_order_validation.
func ValidatePlaceOrder(exchangeSegment, product, price, orderType, quantity, validity, tradingSymbol, transactionType string) error {
	for k, v := range map[string]string{
		"exchange_segment": exchangeSegment, "product": product, "price": price,
		"order_type": orderType, "quantity": quantity, "validity": validity,
		"trading_symbol": tradingSymbol, "transaction_type": transactionType,
	} {
		if err := nonEmpty(v, k); err != nil {
			return err
		}
	}
	if _, ok := ExchangeSegment[exchangeSegment]; !ok {
		return &APIValueError{Msg: "invalid exchange_segment: " + exchangeSegment}
	}
	if _, ok := Product[product]; !ok {
		return &APIValueError{Msg: "invalid product: " + product}
	}
	if _, ok := OrderType[orderType]; !ok {
		return &APIValueError{Msg: "invalid order_type: " + orderType}
	}
	tt := strings.ToUpper(transactionType)
	if tt != "B" && tt != "S" && tt != "BUY" && tt != "SELL" {
		return &APIValueError{Msg: "transaction_type must be B/S"}
	}
	return nil
}

// ValidateCancelOrder — order_id must be non-empty.
func ValidateCancelOrder(orderID string) error {
	return nonEmpty(orderID, "order_id")
}

// ValidateOrderHistory — order_id must be non-empty.
func ValidateOrderHistory(orderID string) error {
	return nonEmpty(orderID, "order_id")
}

// ValidateMargin — basic field presence check.
func ValidateMargin(exchangeSegment, price, orderType, product, quantity, instrumentToken, transactionType string) error {
	for k, v := range map[string]string{
		"exchange_segment": exchangeSegment, "price": price, "order_type": orderType,
		"product": product, "quantity": quantity, "instrument_token": instrumentToken,
		"transaction_type": transactionType,
	} {
		if err := nonEmpty(v, k); err != nil {
			return err
		}
	}
	if _, ok := ExchangeSegment[exchangeSegment]; !ok {
		return &APIValueError{Msg: "invalid exchange_segment: " + exchangeSegment}
	}
	return nil
}

// ValidateLimits — segment/exchange/product against allow-lists.
func ValidateLimits(segment, exchange, product string) error {
	if !contains(SegmentLimits, segment) {
		return &APIValueError{Msg: "invalid segment: " + segment}
	}
	if !contains(ExchangeLimits, exchange) {
		return &APIValueError{Msg: "invalid exchange: " + exchange}
	}
	if !contains(ProductLimits, product) {
		return &APIValueError{Msg: "invalid product: " + product}
	}
	return nil
}

func contains(list []string, v string) bool {
	for _, s := range list {
		if s == v {
			return true
		}
	}
	return false
}
